overlord/snapstate/handlers.go

```diff
@@ -950,13 +950,116 @@ func asyncRefreshOnSnapClose(st *state.State, refreshInfo *userclient.PendingSna
+// trying to launch the snap.
+//
+// This allows the, possibly slow, communication with each snapd session agent,
+// to be performed without holding the snap state lock.
+var asyncBeginDeferredRefreshNotification = func(context context.Context, client *userclient.Client, finishInfo *userclient.FinishedSnapRefreshInfo) {
+	go func() {
+		if err := client.BeginDeferredRefreshNotification(context, finishInfo); err != nil {
+			logger.Noticef("Cannot send notification about finishing deferred refresh: %v", err)
+		}
+	}()
+}
+
+// asyncFinishDeferredRefreshNotification broadcasts desktop notification in a goroutine,
+// telling the user that a refresh process has been completed and that they can now launch
+// the snap.
+//
+// This allows the, possibly slow, communication with each snapd session agent,
+// to be performed without holding the snap state lock.
+var asyncFinishDeferredRefreshNotification = func(context context.Context, client *userclient.Client, finishInfo *userclient.FinishedSnapRefreshInfo) {
+	go func() {
+		if err := client.FinishDeferredRefreshNotification(context, finishInfo); err != nil {
+			logger.Noticef("Cannot send notification about finishing deferred refresh: %v", err)
+		}
+	}()
+}
+
+func calculateTasks(st *state.State, snapName string, instanceName string, revision snap.Revision) (int, string) {
+	remainingTasks := 0
+	newBarText := ""
+	st.Lock()
+	for _, task := range st.Tasks() {
+		if (task.Status() != state.DoStatus) && (task.Status() != state.DoingStatus) {
+			continue
+		}
+		snapsup, _, err := snapSetupAndState(task)
+		if err != nil {
+			continue
+		}
+		if snapsup.SnapName() != snapName {
+			continue
+		}
+		if (snapsup.InstanceName() != instanceName) || (snapsup.Revision() != revision) {
+			continue
+		}
+		if (newBarText == "") && (task.Status() == state.DoingStatus) {
+			newBarText = task.Summary()
+		}
+		remainingTasks++
+	}
+	st.Unlock()
+	return remainingTasks, newBarText
+}
+
+// notifyBeginRefresh uses either snapd-desktop-integration (if available)
+// or the standard desktop notifications to notify the user that a refresh
+// is in progress, and when it has been completed and they can re-launch
+// the application.
+var notifyBeginRefresh = func(st *state.State, taskCount int, instanceName string, revision snap.Revision, snapName string, desktopPath string) {
+	deferredRefreshInfo := userclient.FinishedSnapRefreshInfo{InstanceName: snapName, AppDesktopEntry: desktopPath}
+	clientConnection := userclient.New()
+	asyncBeginDeferredRefreshNotification(context.TODO(), clientConnection, &deferredRefreshInfo)
+
+	clientRefreshStatus := userclient.SnapRefreshProgressInfo{InstanceName: snapName}
+	totalTasks := 0
+	for true {
+		// Wait until the new tasks are added to the list
+		st.Lock()
+		totalTasks = st.TaskCount() - taskCount
+		st.Unlock()
+		if totalTasks != 0 {
+			break
+		}
+		select {
+		case <-time.After(500 * time.Millisecond):
+		}
+	}
+	for true {
```

Comment:
You don't need the 'true'
```suggestion
	for {
```